### PR TITLE
[Api] Visitor and Customer add ProductReview

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Provider/CustomerProvider.php [hunk 1]
```diff
@@ -30,17 +33,33 @@ final class CustomerProvider implements CustomerProviderInterface
...
+        $user = $this->userContext->getUser();
+        if ($user !== null && $user instanceof ShopUserInterface) {
+            return $user->getCustomer();
+        }
+
+        return null;
     }
 
-    public function provide(string $email): CustomerInterface
+    public function getOrCreate(string $email): CustomerInterface
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Provider/CustomerProvider.php:62]: You changed function in service, it should be mentioned in the upgrade file (even if it is `experimental`)

File: src/Sylius/Bundle/ApiBundle/Provider/CustomerProvider.php [hunk 2]
```diff
@@ -30,17 +33,33 @@ final class CustomerProvider implements CustomerProviderInterface
...
+        CustomerRepositoryInterface $customerRepository,
+        UserContextInterface $userContext
     ) {
         $this->canonicalizer = $canonicalizer;
         $this->customerFactory = $customerFactory;
         $this->customerRepository = $customerRepository;
+        $this->userContext = $userContext;
+    }
+
+    public function get(): ?CustomerInterface
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Provider/CustomerProvider.php:51]: I am not sure about this function, firstly this service becomes stateful, secondly, you are mixing two logic (or 3): getting the customer from `userContext`, and getting customer from database (or creating it).

**Review by AdamKasp — COMMENTED** [review]
And when you making a big refactor (not directly related with scop of task), I suggest making it in separate PR, because the majority of this code is a refactor not a feature :)

File: src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     * @var string
+     * @psalm-immutable
+     */
+    public $comment;
+
+    /**
+     * @var string
+     * @psalm-immutable
+     */
+    public $product;
```
[the remark this answers is not in the document]
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php:41]: I'm not sure because all fields in our other endpoints have fileds `product` (or other object fields), and Its mean add `iri`. Our generally idea is adding `iri` if you need to add certain object.

File: src/Sylius/Bundle/ApiBundle/CommandHandler/ChangeShopUserPasswordHandler.php [hunk 4]
```diff
@@ -35,7 +35,7 @@ public function __construct(PasswordUpdaterInterface $passwordUpdater, UserRepos
         $this->userRepository = $userRepository;
     }
 
-    public function __invoke(ChangeShopUserPassword $changeShopUserPassword)
+    public function __invoke(ChangeShopUserPassword $changeShopUserPassword): void
```
[the remark this answers is not in the document]
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/ChangeShopUserPasswordHandler.php:38]: Yep but its **1** line :)

File: src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php (hunk 3 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php:41]:
```suggestion
    public $productCode;
```
as we should use iri only to get a unique identifier for given resource

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        $this->iriConverter = $iriConverter;
+    }
+
+    public function __invoke(AddProductReview $addProductReview): ReviewInterface
+    {
+        /** @var ProductInterface $product */
+        $product = $this->iriConverter->getItemFromIri($addProductReview->product);
+
+        /** @var CustomerInterface|null $customer */
+        $customer = $this->customerProvider->get();
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:59]: I'm not convinced to use this provider in this way here, IMO the command should have information about the email of the current user (if the user is logged in), so it should behave similarly to setting a channel code on commands (in data transformer).

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        $this->customerProvider = $customerProvider;
+        $this->productReviewRepository = $productReviewRepository;
+        $this->productReviewFactory = $productReviewFactory;
+        $this->iriConverter = $iriConverter;
+    }
+
+    public function __invoke(AddProductReview $addProductReview): ReviewInterface
+    {
+        /** @var ProductInterface $product */
+        $product = $this->iriConverter->getItemFromIri($addProductReview->product);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:56]: I guess we should use product repository to get a product by its code instead of using iri in handler, wdyt @lchrusciel?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 7]
```diff
@@ -71,7 +68,7 @@ public function __invoke(PickupCart $pickupCart)
         /** @var ChannelInterface $channel */
         $channel = $this->channelRepository->findOneByCode($pickupCart->getChannelCode());
 
-        $customer = $this->provideCustomer();
+        $customer = $this->customerProvider->get();
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:71]: The same case as with adding a product review, the user shouldn't be provided in handler from context, I would be for reverting this refactor in this PR and do it in a separate PR

File: src/Sylius/Bundle/ApiBundle/Provider/CustomerProvider.php (hunk 1 above)
Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Provider/CustomerProvider.php:62]: I would be for reverting this change

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+use Sylius\Bundle\ApiBundle\Provider\CustomerProviderInterface;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Sylius\Component\Review\Model\ReviewInterface;
+
+final class AddProductReviewHandlerSpec extends ObjectBehavior
+{
+    private $addProductReview;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php:28]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            $productReviewFactory,
+            $iriConverter
+        );
+
+        $this->addProductReview = new AddProductReview(
+            'Good stuff',
+            5,
+            'Really good stuff',
+            'example.com/api/v2/shop/products/666'
+        );
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php:48]: Create a new command in each step:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php:41]: Please, check all command in the cart :), in code we always use some suffix :)

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+        /** @var CustomerInterface|null $customer */
+        $customer = $this->customerProvider->get();
+
+        /** @var string|null $email */
+        $email = $addProductReview->email;
+
+        if ($customer === null && $email === null) {
+            throw new \InvalidArgumentException('Visitor should provide an email');
+        }
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:66]: Btw shouldn't we move this logic to some Validator?

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/command_handlers.xml [hunk 11]
```diff
@@ -30,8 +30,8 @@
             <argument type="service" id="sylius.factory.order" />
             <argument type="service" id="sylius.repository.order" />
             <argument type="service" id="sylius.repository.channel" />
-            <argument type="service" id="sylius.api.context.user" />
             <argument type="service" id="sylius.manager.order" />
+            <argument type="service" id="Sylius\Bundle\ApiBundle\Provider\CustomerProviderInterface"/>
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/command_handlers.xml:34]:
```suggestion
            <argument type="service" id="Sylius\Bundle\ApiBundle\Provider\CustomerProviderInterface" />
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php (hunk 10 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:74]: Yep I add it in next PR that will cover all validation

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/command_handlers.xml (hunk 11 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Resources/config/services/command_handlers.xml]: I add it in new PR with improvements

File: src/Sylius/Behat/Context/Api/Shop/ProductReviewContext.php [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    /** @var ApiClientInterface */
+    private $client;
+
+    /** @var ResponseCheckerInterface */
+    private $responseChecker;
+
+    public function __construct(
+        ApiClientInterface $client,
+        ResponseCheckerInterface $responseChecker
+    ) {
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/ProductReviewContext.php:34]: Could be in one line

File: src/Sylius/Behat/Context/Api/Shop/ProductReviewContext.php [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        $this->client->create();
+    }
+
+    /**
+     * @When I want to review product :product
+     */
+    public function iWantToReviewProduct(ProductInterface $product): void
+    {
+        $this->client->buildCreateRequest();
+        $this->client->addRequestData('productCode', $product->getCode());
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/ProductReviewContext.php:53]: In my previous review, I wrote that we need to keep a product code on command instead of iri, but we still should pass to the endpoint an iri as the resource identifier for API

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+        $product->addReview($review);
+
+        return $review;
+    }
+
+    private function getCustomer(): ?CustomerInterface
+    {
+        /** @var UserInterface|null $user */
+        $user = $this->userContext->getUser();
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:99]: According to my previous comment: https://github.com/Sylius/Sylius/pull/12472#discussion_r600240569 the context shouldn't be used in handler as we would like to maintain stateless behaviour here. As I proposed before, it could behave similarly to setting a channel code on commands (in data transformer).

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        $review->setComment($addProductReview->comment)->shouldBeCalled();
+        $review->setReviewSubject($product->getWrappedObject())->shouldBeCalled();
+        $review->setAuthor($customer)->shouldBeCalled();
+
+        $productReviewRepository->add($review);
+
+        $this($addProductReview);
+    }
+
+    function it_throws_exception_if_shop_user_has_not_been_found(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php:122]:
```suggestion
    function it_throws_an_exception_if_shop_user_has_not_been_found(
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        ShopUserInterface $shopUser,
+        CustomerInterface $customer,
+        ReviewInterface $review
+    ): void {
+        $addProductReview = new AddProductReview(
+            'Good stuff',
+            5,
+            'Really good stuff',
+            'winter_cap'
+        );
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php:61]: Please, don't assign the command to the variable, this makes the test less readable (when I need to search what was the product code for example) and it is not consistent with our style of writing specs.

**Review by GSadee — CHANGES_REQUESTED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml [hunk 17]
```diff
@@ -69,6 +83,9 @@
                 <attribute name="openapi_context">
                     <attribute name="summary">Accepts Product Review</attribute>
                 </attribute>
+                <attribute name="denormalization_context">
+                    <attribute name="groups">admin:product_review:update</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml:87]: we should mention about this change in upgrade file(change from `product_review:update` to `admin:product_review:update`)

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        FactoryInterface $productReviewFactory,
+        ProductRepositoryInterface $productRepository
+    ) {
+        $this->customerProvider = $customerProvider;
+        $this->productReviewRepository = $productReviewRepository;
+        $this->productReviewFactory = $productReviewFactory;
+        $this->productRepository = $productRepository;
+    }
+
+    public function __invoke(AddProductReview $addProductReview): ReviewInterface
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:56]: I'm thinking about this return type. I just checked few handlers and all of them returns void. So is it really needed?

File: src/Sylius/Bundle/ApiBundle/DataTransformer/AddProductReviewInputDataTransformer.php [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+    public function __construct(UserContextInterface $userContext)
+    {
+        $this->userContext = $userContext;
+    }
+
+    public function transform($object, string $to, array $context = [])
+    {
+        /** @var CustomerInterface|null $customer */
+        $customer = $this->getCustomer();
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/DataTransformer/AddProductReviewInputDataTransformer.php:36]: Isn't the $object type of AddProductReview (command)? then we can make a $object->email, and take from User from UserRepository, not with userContext :)
Did I just started to 🤔  stateless? 😄

Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/DataTransformer/AddProductReviewInputDataTransformer.php:36]: This case is for login users, you don't need add email to Request, because it is possible to get email from userContext. `Command` and `CommnadHandler` is stateless, inside theirs all actions are stateless, This operation is doing outside.

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php (hunk 18 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:56]: Command `AddProductReview` create new `ReviewInterface`. Its useful to return `ReviewInterface` as a `Response`.

Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:56]: it is needed if we want to return 201 and created object. But not sure if this is case here

File: src/Sylius/Bundle/ApiBundle/DataTransformer/AddProductReviewInputDataTransformer.php [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+        if ($object->email === null && $customer !== null) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataTransformer/AddProductReviewInputDataTransformer.php:38]:
```suggestion
        if ($customer !== null) {
```

I would say, that we should always override the email (as if you are logged in you can not imposter someone else)

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+        $this(new AddProductReview(
+            'Good stuff',
+            5,
+            'Really good stuff',
+            'winter_cap',
+            'mark@example.com'
+        ));
+    }
+
+    function it_throws_an_exception_if_email_has_not_been_found(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php:76]: we are missing a nulled mail path coverage

File: src/Sylius/Bundle/ApiBundle/spec/DataTransformer/AddProductReviewInputDataTransformerSpec.php [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+
+final class AddProductReviewInputDataTransformerSpec extends ObjectBehavior
+{
+    function let(UserContextInterface $userContext) {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_adds_email_to_add_product_review_for_login_customer(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/DataTransformer/AddProductReviewInputDataTransformerSpec.php:29]:
```suggestion
    function it_adds_email_to_add_product_review_for_logged_in_customer(
```

File: src/Sylius/Bundle/ApiBundle/DataTransformer/AddProductReviewInputDataTransformer.php [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+namespace Sylius\Bundle\ApiBundle\DataTransformer;
+
+use Sylius\Bundle\AdminApiBundle\Model\UserInterface;
+use Sylius\Bundle\ApiBundle\Command\AddProductReview;
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+
+/** @experimental */
+final class AddProductReviewInputDataTransformer implements CommandDataTransformerInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/DataTransformer/AddProductReviewInputDataTransformer.php:23]: I wonder if we should do a more generic data transformer `LoggedInShopUserEmailAwareCommandDataTransformer` similar to `LoggedInShopUserIdAwareCommandDataTransformer`

File: src/Sylius/Bundle/ApiBundle/spec/DataTransformer/AddProductReviewInputDataTransformerSpec.php [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+                'Really good stuff',
+                'winter_cap',
+                'john@example.com'
+            ),
+            'Sylius\Component\Core\Model\ProductReview',
+            []
+        );
+    }
+
+    function it_supports_add_product_review(): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataTransformer/AddProductReviewInputDataTransformerSpec.php:72]:
```suggestion
    function it_supports_command_for_adding_product_review(): void
```
or 
```suggestion
    function it_supports_add_product_review_command(): void
```
?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php (hunk 18 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:56]: If I add void, As a response I get object of `AddProductReview` and `201` code. Its needed? Do user should know structure of this object? This is the intention creating `ReviewInterface` object and it should return this object as a response. IMO

File: src/Sylius/Behat/Context/Api/Shop/ProductReviewContext.php (hunk 13 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/ProductReviewContext.php:78]: Solution for it: https://github.com/Sylius/Sylius/pull/12487

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php (hunk 21 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/spec/CommandHandler/AddProductReviewHandlerSpec.php:76]: Is it not the same case? Is only one way. You must have email, If not its throw exception.

File: UPGRADE-1.9.md [hunk 25]
```diff
@@ -1,3 +1,11 @@
+# UPGRADE FROM `v1.9.2` TO `v1.9.3`
+
+1. To have better control over the serialization process, we introduced `shop` and `admin` prefixes to names of serialization groups on `src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/*` and `src/Sylius/Bundle/ApiBundle/Resources/config/serialization/*`.
+   Several additional serialization groups have been rephrased, to improve readability and predictability of them.
+   If you are using they on your custom entity `api_resource` configuration or serialization groups, you should check if one of these changes may affect on your app. If yes, change all occurs by this pattern:
+
+- `product_review:update` changed to: `admin:product_review:update` and `shop:product_review:update`
+
```
Comment (GSadee) [inline UPGRADE-1.9.md:8]: This note should be in UPGRADE-1.10 file

File: src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+     * @psalm-immutable
+     */
+    public $email;
+
+    public function __construct(
+        string $title,
+        int $rating,
+        string $comment,
+        string $productCode,
+        string $email = null
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/AddProductReview.php:54]:
```suggestion
        ?string $email = null
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+    private $productReviewFactory;
+
+    /** @var ProductRepositoryInterface */
+    private $productRepository;
+
+    public function __construct(
+        CustomerProviderInterface $customerProvider,
+        RepositoryInterface $productReviewRepository,
+        FactoryInterface $productReviewFactory,
+        ProductRepositoryInterface $productRepository
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:48]: I would be for this order:
```suggestion
        FactoryInterface $productReviewFactory,
        RepositoryInterface $productReviewRepository,
        ProductRepositoryInterface $productRepository,
        CustomerProviderInterface $customerProvider
```
here and in variables definitions and assignments

File: src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+        $review = $this->productReviewFactory->createNew();
+        $review->setTitle($addProductReview->title);
+        $review->setRating($addProductReview->rating);
+        $review->setComment($addProductReview->comment);
+        $review->setReviewSubject($product);
+        $review->setAuthor($customer);
+
+        $this->productReviewRepository->add($review);
+
+        $product->addReview($review);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/AddProductReviewHandler.php:81]:
```suggestion
        $product->addReview($review);

        $this->productReviewRepository->add($review);
```

File: src/Sylius/Bundle/ApiBundle/spec/DataTransformer/LoggedInShopUserEmailAwareCommandDataTransformerSpec.php [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+            new AddProductReview(
+                'Good stuff',
+                5,
+                'Really good stuff',
+                'winter_cap'
+            )
+        )->shouldReturn(true);
+    }
+
+    function it_does_not_support_not_add_product_review(): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataTransformer/LoggedInShopUserEmailAwareCommandDataTransformerSpec.php:84]:
```suggestion
    function it_supports_only_add_product_review_command(): void
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Tomasz! :1st_place_medal:
